### PR TITLE
Redmine 4 changes

### PR DESCRIPTION
**Pull request by alexandermeindl**
I did some changes to fix problems with Redmine 4 (hopefully all of it). I also used some rubocop coding standards in changed files.

I comment out quiet_assets gem file, because this do not work with Rails 5. sprockets-rails should do the job with 3.1.0 or newer.

Tests are broken on travis - I suppose we need a own branch for Gemfile changes.

### CODE REVIEW

**Comment by alexandermeindl** [comment]
Does not work in production mode, because https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading-is-disabled-after-booting-in-the-production-environment

Migration for Redmine 4 needs more work.

**Comment by PowerKiKi** [comment]
Did you ever come up with a solution or did you migrate away from redmine_git_hosting ?

**Comment by alexandermeindl** [comment]
I was waiting to this change: https://www.redmine.org/issues/30753, which is required for do more work on this plugin. I'll investigate more time on it in the next weeks, because we need this plugin.

**Comment by PowerKiKi** [comment]
Thank you for your very fast reply. As we are also heavily relying on this plugin we are eager to see this solved.

I am not very fluent in Ruby, but if there's anything I can do to help, let me know.

**Comment by alexandermeindl** [comment]
Rspec tests are totally broken, but application is working with Redmine 4.

**Comment by PowerKiKi** [comment]
Thank you for your work. Do you plan to also fix Travis in the coming days, or should somebody else take care of it ?

**Comment by alexandermeindl** [comment]
I am working on it, but this needs more than a couple of days. Stay tune.

File: app/views/settings/redmine_git_hosting/_gitolite_recycle_bin.html.slim [hunk 1]
```diff
@@ -7,8 +7,9 @@ h3 = l(:label_empty_recycle_bin)
       th = l(:label_recycle_bin_content_size)
       th
         = l(:label_delete_from_recyle_bin)
-        = surround '(', ')' do
```
Comment (n-rodriguez) [inline app/views/settings/redmine_git_hosting/_gitolite_recycle_bin.html.slim]: Hi Alexander! Thanks for your work :+1: 
Just for your information this is pure HAML helper : http://haml.info/docs/yardoc/file.REFERENCE.html#surround

Comment (n-rodriguez) [reply app/views/settings/redmine_git_hosting/_gitolite_recycle_bin.html.slim]: Ok... I've just seen you switched to slim ^^

**Comment by alexandermeindl** [comment]
Hi Nicolas,

yes, I switched to Slim. Other big chanages are drop redmine_bootstrap_kit dependency and use [additionals plugin](https://github.com/AlphaNodes/additionals), which is maintained permanently by our company. Last big change is the switch to a flat translation file structure, to make translation more easy for non-developer.
At this point now everythings looks quit good. Missing TODOs: fixing the 53 tests, which fails. Most of them are mass assignment problems. For there we need another integration, see https://github.com/thoughtbot/shoulda-matchers/issues/243

I hope all changes are acceptable (spent quit a lot time for it) and maybe this pull request is ready to merge right now - so we can work together on fine tuning and tests.

**Comment by Alwaysin** [comment]
Hi @alexandermeindl and thanks for your work!
I wasn't able to try your updated version (taken from cloning https://github.com/AlphaNodes/redmine_git_hosting) because I got this:

```
rake aborted!
LocalJumpError: unexpected return
/usr/src/redmine/plugins/redmine_git_hosting/lib/tasks/ci_tools.rake:3:in `block (2 levels) in <top (required)>'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/task_manager.rb:224:in `in_namespace'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/dsl_definition.rb:141:in `namespace'
/usr/src/redmine/plugins/redmine_git_hosting/lib/tasks/ci_tools.rake:2:in `block in <top (required)>'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/task_manager.rb:224:in `in_namespace'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/dsl_definition.rb:141:in `namespace'
/usr/src/redmine/plugins/redmine_git_hosting/lib/tasks/ci_tools.rake:1:in `<top (required)>'
/usr/src/redmine/lib/tasks/redmine.rake:190:in `load'
/usr/src/redmine/lib/tasks/redmine.rake:190:in `block in <top (required)>'
/usr/src/redmine/lib/tasks/redmine.rake:190:in `each'
/usr/src/redmine/lib/tasks/redmine.rake:190:in `<top (required)>'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/engine.rb:650:in `load'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/engine.rb:650:in `block in run_tasks_blocks'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/engine.rb:650:in `each'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/engine.rb:650:in `run_tasks_blocks'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/application.rb:515:in `run_tasks_blocks'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/engine.rb:459:in `load_tasks'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/railtie.rb:190:in `public_send'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/railties-5.2.3/lib/rails/railtie.rb:190:in `method_missing'
/usr/src/redmine/Rakefile:7:in `<top (required)>'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/rake_module.rb:29:in `load'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/rake_module.rb:29:in `load_rakefile'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:703:in `raw_load_rakefile'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:104:in `block in load_rakefile'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:103:in `load_rakefile'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:82:in `block in run'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/lib/rake/application.rb:80:in `run'
/usr/src/redmine/vendor/cache/ruby/2.3.0/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
/usr/src/redmine/vendor/cache/ruby/2.3.0/bin/rake:22:in `load'
/usr/src/redmine/vendor/cache/ruby/2.3.0/bin/rake:22:in `<main>'
```

Do you have any clue why it fails?

**Comment by alexandermeindl** [comment]
Hi Alwaysin,
I did some changes with my last commit. Can you try it again?

**Comment by Alwaysin** [comment]
It's much better now, it goes through! Thanks!

**Comment by PowerKiKi** [comment]
@Alwaysin are you using this PR in production ? how is it working for you ?

@n-rodriguez, you did not reply about merging this PR. Would you agree to merge it and then accept a second PR for tests ?

File: lib/redmine_git_hosting/patches/repositories_controller_patch.rb [hunk 2]
```diff
@@ -141,12 +141,18 @@ def diff_with_options
                           Digest::MD5.hexdigest("#{@path}-#{@rev}-#{@rev_to}-#{@diff_type}-#{current_language}")
           unless read_fragment(@cache_key)
             @diff = @repository.diff(@path, @rev, @rev_to)
+            unless @diff
+              show_error_not_found
+              return
+            end
+
             show_error_not_found unless @diff
```
Comment (n-rodriguez) [inline lib/redmine_git_hosting/patches/repositories_controller_patch.rb]: There is a duplication here. I think just adding `return` in front of  `show_error_not_found unless @diff` should work.

Comment (alexandermeindl) [reply lib/redmine_git_hosting/patches/repositories_controller_patch.rb]: Hi Nicolas, you are right, this was duplicated, I removed it with my last commit.

**Comment by n-rodriguez** [comment]
Great job :+1:

**Comment by PowerKiKi** [comment]
@n-rodriguez will you have some availability yourself in the coming days to help with pushing a new release including this awesome PR ?

**Comment by n-rodriguez** [comment]
sure!

**Pull request merged**